### PR TITLE
Queue newly created and updated annotations for syncing to Elasticsearch

### PR DESCRIPTION
**Pull request by seanh**

### CODE REVIEW

File: tests/h/storage_test.py [hunk 1]
```diff
@@ -92,7 +94,7 @@ def test_expand_uri_document_uris(self, db_session):
         ]
 
 
-@pytest.mark.usefixtures("models", "groupfinder_service", "update_document_metadata")
+@pytest.mark.usefixtures("models")
```
Comment (seanh) [inline tests/h/storage_test.py:97]: Removed two unnecessary `usefixtures`'s here

File: tests/h/storage_test.py [hunk 2]
```diff
@@ -391,7 +404,7 @@ def annotation_data(self):
         }
 
 
-@pytest.mark.usefixtures("models", "update_document_metadata")
+@pytest.mark.usefixtures("models")
```
Comment (seanh) [inline tests/h/storage_test.py:407]: Removed an unnecessary `usefixtures` here

File: tests/h/storage_test.py [hunk 3]
```diff
@@ -656,7 +680,7 @@ def models(patch):
     return models
 
 
-@pytest.fixture
+@pytest.fixture(autouse=True)
```
Comment (seanh) [inline tests/h/storage_test.py:683]: Made this module-level patch fixture an autouse as per usual

File: h/storage.py [hunk 4]
```diff
@@ -147,6 +147,10 @@ def create_annotation(request, data, group_service):
     request.db.add(annotation)
     request.db.flush()
 
+    request.find_service(name="search_index")._queue.add_by_id(
+        annotation.id, tag="storage.create_annotation", schedule_in=30
+    )
```
Comment (jon-betts) [inline h/storage.py:152]: I don't think this is the right place to do it. Currently:

 * In `h.views.api.annotations.create()` we call `storage.create_annotation`
 * Then we call `_publish_annotation_event` which creates `AnnotationEvent`
 * This is picked up by `h.subscribers.annotation_sync()`
 * This calls `h.services.search_index.service.handle_annotation_event()`
 * This is where we do primary indexing (or async fallback if that fails)

I think this last point is where any addition 3rd layer back up should take place.

**Review by jon-betts — CHANGES_REQUESTED** [review]
Sorry, this isn't being done in the right place. There's more details below, but basically this should happen in `SearchIndex.handle_annotation_event()` where the other single indexing parts happen.

File: h/storage.py (hunk 4 above)
Comment (seanh) [reply h/storage.py:152]: This can't be done in that event subscriber because this has to happen within the request's original DB transaction: the atomic transactionality (so it definitely can't create/update an annotation in the DB without also adding the job) is a crucial aspect.

Comment (seanh) [reply h/storage.py:152]: @jon-betts Do you happen to know what the upper time limit is on how long the "synchronous" annotation indexing can take, if it uses the maximum number of Celery retries? We should set the `schedule_in=30` to be comfortably longer than that

**Comment by jon-betts** [comment]
We've had a discussion about this, some highlights. There are two ways (proposed) of doing this which have different properties:

1. Do the queue in the same transaction as the annotation add (this PR)

 * Makes the chance of accepting an annotation but not syncing it close to zero
 * It does this at the cost of increasing the chance we will reject an annotation
 * When the service says it's accepted an annotation it _should_ get synced eventually

2. Use the second transaction which is used for the sync and delayed sync fallback

 * Does not increase the chance of rejecting an annotation
 * It does this at the cost of increasing the chance we will fail to sync
 * When the service says it's accepted an annotation it _might not_ get synced eventually

The question of which is better then largely comes down to:

 * How likely is it we would fail in the first transaction because of the additional work vs...
 * How likely is it we would fail because we start the work in a second transaction?
    * Sean's position is the first is essentially zero, and the second is significantly higher
    * I don't see why, and I suspect they are similarly reliable (so it's a trade)
    * I also think there are scenarios where an outright success could be converted into a failure
 * The second is is it better to reject data outright, or accept it and leave the possibility of fixing it up later
   * Sean's position is that certainty about the outcome trumps total % success
   * As long as the % success rate difference is very small I agree (if it was a difference of 50%, it's another matter)
   * I think we both agree the difference is likely to be tiny

## What's the trade off here then?

 * This PR makes it the same, or less likely we will accept annotations
 * It makes it the same or more likely that when we accept an annotation we will really sync it
 * So our precision has potentially increased ("when we say we worked, you can rely on it")
 * But our success rate has potentially decreased ("we say we worked less often")
 * The differences are unknown and likely tiny
 * The differences are not going to be the same (we hope we have greatly increased our precision at a tiny cost to our success rate)

This is probably a good trade off, as users can now get direct feedback when we don't accept an annotation and try again.

**Review by jon-betts — CHANGES_REQUESTED** [review]
We've discussed larger issues about the trade offs in this PR. The approach here is likely better at the costs discussed in the comments.

Aside from that there are some things I think could be improved.

 * We now have two different places that kick of indexing (and emails etc.): the storage, and the view
 * It seems to me one or the other should be in charge, but not both
 * I think we should have outward facing methods on `search_index` service for doing whatever you want to do rather than access the queue
 * I'd like the existence of the queue to remain secret so we can change our mind without rewriting tests and classes everywhere

**Comment by jon-betts** [comment]
I'm aware this is getting merged without addressing the fixes suggested above. We may come back to some of these or not.

File: h/storage.py (hunk 4 above)
Comment (jon-betts) [reply h/storage.py:152]: The synchronous indexing doesn't involve rabbit at all, so there shouldn't be any retries. It should be the total time we would take to obtain an ES handle from a pool I assume (which may have retries) and perform the update.

I don't know how long that is, but the max timeouts I've seen in NewRelic for ES calls are about 20s, so 30s would be safe.

Comment (seanh) [reply h/storage.py:152]: It [does send the task to Rabbit it Elasticsearch fails](https://github.com/hypothesis/h/blob/5f207cd5ca428ae915bbe08196417ff39d52a429/h/services/search_index/service.py#L130), right? I think in the worse case scenario where Elasticsearch fails and the maximum amount of Celery retries get used, we don't want the job queue to kick in until _after_ all of those retries

Comment (jon-betts) [reply h/storage.py:152]: Clarification from slack, but the "synchronous" here we mean the initial attempt in process + the delayed celery attempts.

The synchronous indexing doesn't involve rabbit at all, so there shouldn't be any retries. It should be the total time we would take to obtain an ES handle from a pool I assume (which may have retries) and perform the update.

I don't know how long that is, but the max timeouts I've seen in NewRelic for ES calls are about 20s, so 30s would be safe (assuming this includes the time to get a handle). If the handle has a separate timeout, it could be longer.

The celery task has a similar situation obtaining a handle from the pool, but the two outcomes here are pretty much:

* It gets one right away
* It totally fails (presumably at the max timeout)

The timeout to get a handle is less than 2s by `RETRY_POLICY_QUICK`) but assuming that works and we then go onto fail repeatedly while retrying, the second scenario would probably be longer and dominate.

We use exponential back off and jitter up to 5 times. So without jitter I think it's 1 + 2 + 4 + 8 + 16 (32), but jitter treats all of these values as a maximum, so an average of 16s max 32s, ignoring time to do anything.

So in summary in the worst case

 * However long it takes to get an ES handle
 * + however long it takes ES to timeout
 * + How long it takes to timeout connecting to Rabbit (<2s)
 * + 32s (retry delays)
 * + any time spent in task doing things and failing

So I'd say more minutes than seconds.

Comment (seanh) [reply h/storage.py:152]: I think the Elasticsearch timeout is actually 5 seconds. But I think to be on the safe side we could `schedule_in=180` (three mins), that seems fine to me
